### PR TITLE
128-EEA3 implementation

### PR DESCRIPTION
**Pull request by jackyliu16**
implementation EEA-128 algorithm, based on [GB/T 33133.2-2021](https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3)

### CODE REVIEW

**Comment by Nugine** [comment]
CC @Xerxes-zhz

**Comment by Nugine** [comment]
可参考 GmSSL 的实现，对比每一步，看哪里有问题

入口点
+ https://github.com/guanzhi/GmSSL/blob/34fa519dc0f94a9a3995d9daf09c84cdac37abd8/src/zuc_modes.c#L32-L49

**Comment by jackyliu16** [comment]
> 可参考 GmSSL 的实现，对比每一步，看哪里有问题
> 
> 入口点
> 
>     * https://github.com/guanzhi/GmSSL/blob/34fa519dc0f94a9a3995d9daf09c84cdac37abd8/src/zuc_modes.c#L32-L49
> 
> 
> gm-zuc
> 
>     * https://github.com/CrayfishGo/gm-rs/blob/main/gm-zuc/src/eea.rs

好的,我去研究一下
目前在跟这个对比  git@github.com:guyongqiangx/crypto-work.git (push)

File: Cargo.toml [hunk 1]
```diff
@@ -15,4 +15,4 @@ all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
-const-str = "0.5.7"
+const-str = "0.5.7"
```
Comment (Nugine) [inline Cargo.toml:18]: 这里没必要删掉换行符

File: src/eea3_128.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
+#![allow(unused_imports, unused_variables)]
```
Comment (Nugine) [inline src/eea3_128.rs:1]: unused 相关 lints 需解决，不能 allow

File: src/zuc128.rs [hunk 3]
```diff
@@ -1,67 +1,8 @@
 //! ZUC-128 Algorithms
 
+use crate::consts::{D, S0, S1};
```
Comment (Nugine) [inline src/zuc128.rs:3]: D,S0,S1 属于 ZUC128 标准的一部分，没必要提取到单独模块，其他模块要引用相同常量时可设置为 pub(crate)

File: src/eea3_128.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+///
+/// # input:
+/// - ck:       128bit  confidentiality key
+/// - iv:       128bit  initial vector
+/// - length:       32bit   bit length of plaintext information stream
+/// - ibs:          &[u8]   input bitstream
+///
+/// # output:
+/// - Vec<u8>:  encrypted bit stream
+#[allow(dead_code)]
```
Comment (Nugine) [inline src/eea3_128.rs:17]: dead_code 需解决，不能 allow

File: src/eea3_128.rs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+#[cfg(test)]
+mod tests {
+    use crate::eea3_128::eea3_128;
+    use crate::ZUC128;
+    use std::fmt::Debug;
+
+    #[test]
+    fn test_1() {
+        // 3GPP LTE Example 1
```
Comment (Nugine) [inline src/eea3_128.rs:85]: 最好标注测试用例来源，包括网址、章节。

File: src/lib.rs [hunk 6]
```diff
@@ -19,5 +19,10 @@
 // ---
 #![cfg_attr(docsrs, feature(doc_cfg))]
 
+mod consts;
+
+/// Confidentiality Algorithm
```
Comment (Nugine) [inline src/lib.rs:24]: 模块文档通常以 `//!` 内部属性的形式写在对应模块文件的开头。

File: src/eea3_128.rs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+/// - count:        32bit   counter
+/// - bearer:       8bit    carrier layer identification
+/// - direction:    1bit    transmission direction identification
+/// - ck:           128bit  confidentiality key
+/// - length:       32bit   bit length of plaintext information stream
+/// - ibs:          &[u8]   input bitstream
+///
+/// # output:
+/// - Vec<u8>:  encrypted bit stream
+#[allow(dead_code)]
```
Comment (Nugine) [inline src/eea3_128.rs:57]: dead_code 需解决，不能 allow

**Review by Nugine — CHANGES_REQUESTED** [review]
主要建议
+ review 中提出的修改要求
+ 尝试减少 eea3_128 的内存分配次数，最好仅需单次 Vec 分配，无需反复扩容和多次创建

@Xerxes-zhz 若 #4 先合并的话可能与 #3 产生冲突，你也看一下有没有其他问题

File: src/eea3_128.rs (hunk 4 above)
Comment (jackyliu16) [reply src/eea3_128.rs:17]: 这个我感觉不是目前我需要实现的部分?
这个地方是在 clippy 那边检测出来的问题, 报错当前函数未被调用, 但是这个理论上不是应该由使用者调用的么?

File: src/zuc128.rs (hunk 3 above)
Comment (jackyliu16) [reply src/zuc128.rs:3]: 这个地方我检查过后面有关于eia3算法的实现, 没有进一步引用这部分的常量的必要
感觉这样会使整体代码更加美观?  (毕竟将一些与实现无关的部分分离出去了)

File: src/eea3_128.rs (hunk 5 above)
Comment (Nugine) [reply src/eea3_128.rs:85]: https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf
这里还有 5 个 test set，最好也加上

File: src/zuc128.rs (hunk 3 above)
Comment (Nugine) [reply src/zuc128.rs:3]: 我个人偏好：属于同一逻辑且无需复用的，尽量放在同一个地方，少拆文件

File: src/eea3_128.rs (hunk 4 above)
Comment (Nugine) [reply src/eea3_128.rs:17]: 公开给用户的 API 设置为 pub，这样就不会报未调用了

Comment (jackyliu16) [reply src/eea3_128.rs:17]: 笑死,才发现我不知道啥时候把pub搞没了
应该是某段时间rebase搞炸了弄的

Comment (jackyliu16) [reply src/eea3_128.rs]: 这个怎么处理呢?

```    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.22s
just test
cargo test --all-features
   Compiling zuc v0.1.0 (/home/jacky/Coding/zuc)
warning: unused import: `std::ops::BitXor`
 --> src/eea3_128.rs:5:5
  |
5 | use std::ops::BitXor;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: function `encryption_xor` is never used
  --> src/eea3_128.rs:19:8
   |
19 | pub fn encryption_xor(ck: u128, iv: u128, length: u32, ibs: &[u8]) -> Vec<u8> {
   |        ^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: function `eea3_128` is never used
  --> src/eea3_128.rs:58:8
   |
58 | pub fn eea3_128(
   |        ^^^^^^^^

warning: `zuc` (lib) generated 3 warnings (run `cargo fix --lib -p zuc` to apply 1 suggestion)
warning: `zuc` (lib test) generated 1 warning (1 duplicate)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.19s
     Running unittests src/lib.rs (target/debug/deps/zuc-72289d4287fbb12b)

running 6 tests
test eea3_128::tests::test_1 ... ok
test eea3_128::tests::test_2 ... ok
test eea3_128::tests::test_set_3_from_gmssl ... ok
test zuc128::tests::examples ... ok
test eea3_128::tests::test_set_4_from_gmssl ... ok
test eea3_128::tests::test_3 ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests zuc

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
                                                                                                                                                                                                    /0.6s
➜  zuc git:(main) ✗ ga .                                                                                                                                                                            /0.0s
➜  zuc git:(main) ✗ gc  
[main 81d90e8] tests(eea-128): add test case from gmssl
 1 file changed, 110 insertions(+), 7 deletions(-)                                                                                                                                                 /17.2s
➜  zuc git:(main) just ci 
cargo fmt --all --check
cargo clippy --all-features -- -D warnings
    Checking zuc v0.1.0 (/home/jacky/Coding/zuc)
error: unused import: `std::ops::BitXor`
 --> src/eea3_128.rs:5:5
  |
5 | use std::ops::BitXor;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: function `encryption_xor` is never used
  --> src/eea3_128.rs:19:8
   |
19 | pub fn encryption_xor(ck: u128, iv: u128, length: u32, ibs: &[u8]) -> Vec<u8> {
   |        ^^^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`

error: function `eea3_128` is never used
  --> src/eea3_128.rs:58:8
   |
58 | pub fn eea3_128(
   |        ^^^^^^^^

error: could not compile `zuc` (lib) due to 3 previous errors
error: Recipe `ci` failed on line 20 with exit code 101                                                                                                                                             /0.3s
➜  zuc git:(main) 
```

如果在lib中添加对于方法的使用,就会报错  this function could have a `#[must_use]` attribute

Comment (Nugine) [reply src/eea3_128.rs]: 公开给用户的 api 还需要 re-export 或 pub mod

https://github.com/Nugine/zuc/blob/6058a9df895170fac4bdb5b8edeb21dfd187319f/src/lib.rs#L22-L23

`just doc` 打开文档，看下用户视角

Comment (Nugine) [reply src/eea3_128.rs]: `#[must_use]` 的 lint 一般没问题，防止计算结果被意外丢弃，引发逻辑错误

Comment (jackyliu16) [reply src/eea3_128.rs]: 我这个后面那句话有说

> 公开给用户的 api 还需要 re-export 或 pub mod
> 
> https://github.com/Nugine/zuc/blob/6058a9df895170fac4bdb5b8edeb21dfd187319f/src/lib.rs#L22-L23
> 
> `just doc` 打开文档，看下用户视角

我这个后面那句话有说,  最新的 commit 修复了这个问题

File: src/eea3_128.rs [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            .iter()
+            .flat_map(|&word| word.to_be_bytes())
+            .collect::<Vec<u8>>();
+        assert_eq!(eea3_128(count, bearer, direction, ck, length, &ibs), obs);
+    }
+
+    /// Test Set 3 from gsma
+    /// https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf
+    #[test]
+    fn test_set_4_from_gmssl() {
```
Comment (Nugine) [inline src/eea3_128.rs:256]:
```suggestion
    /// Test Set 4 from gsma
    /// https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf
    #[test]
    fn test_set_4_from_gsma() {
```

File: src/eea3_128.rs [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            .iter()
+            .flat_map(|&word| word.to_be_bytes())
+            .collect::<Vec<u8>>();
+        assert_eq!(eea3_128(count, bearer, direction, ck, length, &ibs), obs);
+    }
+
+    /// Test Set 3 from gsma
+    /// https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf
+    #[test]
+    fn test_set_3_from_gmssl() {
```
Comment (Nugine) [inline src/eea3_128.rs:213]:
```suggestion
    /// Test Set 3 from gsma
    /// https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf
    #[test]
    fn test_set_3_from_gsma() {
```

File: src/eea3_128.rs [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+            0x50, 0xfc, 0x0f, 0x6f, 0x26, 0x13, 0xb2, 0xf2, 0xbc, 0xa6, 0xdf, 0x5a, 0x47, 0x3a,
+            0x57, 0xa4, 0xa0, 0x0d, 0x98, 0x5e, 0xba, 0xd8, 0x80, 0xd6, 0xf2, 0x38, 0x64, 0xa0,
+            0x7b, 0x01,
+        ];
+
+        assert_eq!(eea3_128(count, bearer, direction, ck, length, &ibs), &obs);
+    }
+
+    /// 3GPP LTE Example 3
+    /// FROM https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3
```
Comment (Nugine) [inline src/eea3_128.rs:148]: GSMA 与 GB/T 33133.2-2021 重复的测试用例可以标上两个来源地址

File: src/eea3_128.rs [hunk 11]
```diff
@@ -0,0 +1,305 @@
+//! ZUC Confidentiality Algorithm
+//! ([GB/T 33133.1-2016](https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3 ))
```
Comment (Nugine) [inline src/eea3_128.rs:2]: 链接名称与 URL 不一致

File: src/eea3_128.rs [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+
+use crate::ZUC128;
+
+/// zuc xor encryption algorithm
+/// ([GB/T 33133.1-2016](https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3 ))
```
Comment (Nugine) [inline src/eea3_128.rs:7]: 链接名称与 URL 不一致

File: src/eea3_128.rs [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+        keys[i] = 0x0;
+    }
+
+    let mut res = ibs.to_vec();
+    res.iter_mut().zip(keys.iter()).for_each(|(ib, k)| *ib ^= k);
+    res
+}
+
+/// eea3-128 privacy algorithm (3GPP LTE)
+/// ([GB/T 33133.1-2016](https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3 ))
```
Comment (Nugine) [inline src/eea3_128.rs:47]: 应指向 128-EEA3 的标准文件。
文档中正式名称应为 128-EEA3。

File: src/eea3_128.rs [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+        ];
+
+        assert_eq!(
+            eea3_128(count, bearer, direction, ck, length, &ibs),
+            &obs[..]
+        )
+    }
+
+    /// 3GPP LTE Example 2
+    /// FROM https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3
```
Comment (Nugine) [inline src/eea3_128.rs:115]: GSMA 与 GB/T 33133.2-2021 重复的测试用例可以标上两个来源地址

File: src/eea3_128.rs [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+
+    encryption_xor(ck, u128::from_ne_bytes(iv), length, ibs)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::eea3_128::eea3_128;
+
+    /// 3GPP LTE Example 1
+    /// FROM https://openstd.samr.gov.cn/bzgk/gb/newGbInfo?hcno=5D3CBA3ADEC7989344BD1E63006EF2B3
```
Comment (Nugine) [inline src/eea3_128.rs:91]: GSMA 与 GB/T 33133.2-2021 重复的测试用例可以标上两个来源地址

File: src/eea3_128.rs [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+    iv[0] = (count >> 24) as u8;
+    iv[1] = (count >> 16) as u8;
+    iv[2] = (count >> 8) as u8;
+    iv[3] = count as u8;
+    iv[4] = bearer << 3 | direction << 2;
+    iv[5..=7].fill(0x0);
+    let tmp = &iv[0..8].to_vec();
+    iv[8..16].copy_from_slice(tmp);
+
+    encryption_xor(ck, u128::from_ne_bytes(iv), length, ibs)
```
Comment (Nugine) [inline src/eea3_128.rs:83]: 存在多于一次内存分配，最好优化到单次分配

File: src/eea3_128.rs [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+    if length % 8 != 0 {
+        keys[length as usize / 8] &= 0xFF << (8 - length % 8);
+    }
+    for i in length as usize / 8 + 1..keys.len() {
+        keys[i] = 0x0;
+    }
+
+    let mut res = ibs.to_vec();
+    res.iter_mut().zip(keys.iter()).for_each(|(ib, k)| *ib ^= k);
+    res
```
Comment (Nugine) [inline src/eea3_128.rs:43]: 存在多于一次内存分配，最好优化到单次分配

File: src/eea3_128.rs (hunk 13 above)
Comment (jackyliu16) [reply src/eea3_128.rs:47]: emm, 你指的是这个么?
[Specification of the 3GPP Confidentiality and
Integrity Algorithms 128-EEA3 & 128-EIA3.
Document 2: ZUC Specification?](https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/EEA3_EIA3_specification_v1_8.pdf)

File: src/eea3_128.rs (hunk 16 above)
Comment (jackyliu16) [reply src/eea3_128.rs:83]: 我晚上看一下有没有办法解决这个可变引用和不可变引用同时存在的问题, 实在不行直接重复写 (

File: src/eea3_128.rs (hunk 13 above)
Comment (Nugine) [reply src/eea3_128.rs:47]: 这个，关于 128-EEA3 的 spec
https://www.gsma.com/solutions-and-impact/technologies/security/wp-content/uploads/2019/05/EEA3_EIA3_specification_v1_8.pdf

File: src/eea3_128.rs [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+    let mut zuc = ZUC128::new(&ck, &iv);
+    let mut res = (0..l)
+        .flat_map(|_| {
+            let z = zuc.generate().to_be();
+            [z as u8, (z >> 8) as u8, (z >> 16) as u8, (z >> 24) as u8]
+        })
+        .enumerate()
+        .map(|(i, byte)| byte ^ ibs[i])
+        .collect::<Vec<u8>>();
```
Comment (Nugine) [inline src/eea3_128.rs:34]: 用预分配方法更好
1. 复制 ibs 到 Vec，作为 res
2. 根据密钥原地修改 res 的内容
3. 末尾处理

**Review by Nugine — CHANGES_REQUESTED** [review]
剩余工作
+ rebase 到 main，并修复 lint 报错
+ 解决 review 要求

完成之后，若准备好合并，发一条 `Ready to merge` 消息即可

File: src/eea3_128.rs [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+        res[length as usize / 8] |= ibs[length as usize / 8] & (0xFF >> (8 - length % 8));
+    }
+    for i in length as usize / 8 + 1..res.len() {
+        res[i] = 0x0;
+    }
+
+    res
+}
+
+/// eea3-128 privacy algorithm (3GPP LTE)
```
Comment (Nugine) [inline src/eea3_128.rs:47]:
```suggestion
/// 128-EEA3: 3GPP confidentiality algorithm
```

File: Cargo.toml [hunk 20]
```diff
@@ -16,3 +16,4 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
 const-str = "0.5.7"
+
```
[the remark this answers is not in the document]
Comment (jackyliu16) [reply Cargo.toml:19]: 这个地方我有点不知道怎么删除对于这个文件的修改（

[the remark this answers is not in the document]
Comment (Nugine) [reply Cargo.toml:19]: 可以加个删掉空行的新提交

**Review by Nugine — CHANGES_REQUESTED** [review]
剩余任务：
+ 看 CI lint 报错，修复 tests 中相关格式问题
+ 解决 review 要求

我们希望代码质量尽可能做到最优，很多高要求的开源项目 PR 都会经过反复多轮修改，长review 周期很常见。这也是一次练习 Rust 的机会，慢慢来。

File: src/eea3_128.rs [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+/// - iv:       128bit  initial vector
+/// - length:   32bit   bit length of plaintext information stream
+/// - ibs:      &[u8]   input bitstream
+///
+/// # output:
+/// - Vec<u8>:  encrypted bit stream
+#[must_use]
+#[allow(clippy::needless_range_loop)]
+#[allow(clippy::cast_possible_truncation)]
+pub fn encryption_xor(ck: u128, iv: u128, length: u32, ibs: &[u8]) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (jackyliu16) [reply src/eea3_128.rs:20]: 这个与算法实现的不一致了吧,这样操作是可以的么?
标准里面有明确说明对应的端绪是怎样的
image

**Comment by jackyliu16** [comment]
有点难绷了, 你这一个接一个的, 然后这 clippy 天天莫名其妙的问题
要我 0x33532455  改成 0x0x3353_2455 就算了, 然后过长还给我自动格式化成单列然后报错函数太长
我自己处理 nixos 的 review 都没那么多事情.

File: src/eea3_128.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Nugine) [reply src/eea3_128.rs]: 可以的，`&[u8;16]` 和 `u128` 都是 128bit，前者表意更明确，实际使用也更方便

**Comment by Nugine** [comment]
抱歉。这次活动是我第一次试验带奖励的 PR 模式，对代码质量的期待可能过高，后续将会调整。
对于不合理的 lint 要求，我将自己尝试后再提出，或者直接自行修改。
对于代码逻辑和文档的要求，应更清晰明确，给出必要的提示。

由于您未能达成单次内存分配的 review 要求，最终奖励将有扣减。
此次 PR 我已自行修改到符合要求。如果您仍对其他 todo item 感兴趣，欢迎继续参与奖励活动。

**Pull request merged**